File: src/lib/d1.ts
/** Small D1 helpers — all dashboard SQL lives in api/*.ts */
export type D1Row = Record<string, unknown>;

export async function first<T extends D1Row>(
  db: D1Database,
  sql: string,
  ...bindings: unknown[]
): Promise<T | null> {
  const stmt = db.prepare(sql);
  const bound = bindings.length ? stmt.bind(...bindings) : stmt;
  return (await bound.first<T>()) ?? null;
}

export async function all<T extends D1Row>(
  db: D1Database,
  sql: string,
  ...bindings: unknown[]
): Promise<T[]> {
  const stmt = db.prepare(sql);
  const bound = bindings.length ? stmt.bind(...bindings) : stmt;
  const { results } = await bound.all<T>();
  return results ?? [];
}

File: src/lib/d1.py
def row_val(row, key, default=None):
    if row is None:
        return default
    try:
        v = row[key]
        return default if v is None else v
    except (KeyError, TypeError, IndexError):
        getter = getattr(row, "get", None)
        if callable(getter):
            return getter(key, default)
        return default


def to_plain(value):
    if value is None:
        return None
    if isinstance(value, (str, int, float, bool)):
        return value
    keys_fn = getattr(value, "keys", None)
    if callable(keys_fn):
        return {str(k): to_plain(value[k]) for k in keys_fn()}
    try:
        return [to_plain(item) for item in value]
    except TypeError:
        return str(value)


async def first(db, sql, *bindings):
    stmt = db.prepare(sql)
    if bindings:
        stmt = stmt.bind(*bindings)
    row = await stmt.first()
    return to_plain(row)


async def all_rows(db, sql, *bindings):
    stmt = db.prepare(sql)
    if bindings:
        stmt = stmt.bind(*bindings)
    result = await stmt.all()
    if not result:
        return []
    rows = getattr(result, "results", None) or []
    return [to_plain(row) for row in rows]
